common/oe_host_socket.h

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#ifndef _OE_HOST_SOCKET_H
+#define _OE_HOST_SOCKET_H
+
+#include <openenclave/corelibc/bits/types.h>
+#include <openenclave/internal/syscall/types.h>
+#include <stdint.h>
+
+#ifdef _WIN32
+#include <winsock2.h>
+#else
+#include <sys/socket.h>
+#endif
+
+#define GETADDRINFO_HANDLE_MAGIC 0xed11d13a
+
+typedef struct _getaddrinfo_handle
+{
+    uint32_t magic;
+    struct addrinfo* res;
+    struct addrinfo* next;
+} getaddrinfo_handle_t;
+
+OE_INLINE getaddrinfo_handle_t* _cast_getaddrinfo_handle(void* handle_)
+{
+    getaddrinfo_handle_t* handle = (getaddrinfo_handle_t*)handle_;
+
+    if (!handle || handle->magic != GETADDRINFO_HANDLE_MAGIC || !handle->res)
+        return NULL;
+
+    return handle;
+}
+
+OE_INLINE int _getaddrinfo_read(
+    uint64_t handle_,
+    int* ai_flags,
+    int* ai_family,
+    int* ai_socktype,
+    int* ai_protocol,
+    oe_socklen_t ai_addrlen_in,
+    oe_socklen_t* ai_addrlen,
+    struct oe_sockaddr* ai_addr,
+    size_t ai_canonnamelen_in,
+    size_t* ai_canonnamelen,
+    char* ai_canonname,
+    int* err_no)
+{
+    int ret = -1;
+    getaddrinfo_handle_t* handle = _cast_getaddrinfo_handle((void*)handle_);
+
+    if (!err_no)
+    {
+        goto done;
+    }
+
+    if (!handle || !ai_flags || !ai_family || !ai_socktype || !ai_protocol ||
+        !ai_addrlen || !ai_canonnamelen)
+    {
+        *err_no = OE_EINVAL;
```

Comment:
Does posix mandate that in these scenarios, EINVAL code must be returned. It would be good to document why we are returning specific error codes.

Reply:
getaddrinfo_read isn't a real syscall, its just used as a helper to allow the enclave to get the next record. I mostly just took the existing code from host/linux/syscall.c and moved it here so it could be reused.

In general, EINVAL means invalid argument. If the caller passes in null pointers, I think this is that right error to return.

Reply:
Since there is the concept of `next record`, is this function thread-safe?

Reply:
It isn't. The handle passed in is retrieved with getaddrinfo_open ocall and a new structure is allocated for each open. So if two threads call open, their read calls will be on different structures.

If one thread calls open and then two threads call read on that same handle, then yes there is a race condition.

getaddrinfo is only exposed to the enclave as an atomic operation which calls open, read, and free, then returns the data to the caller. So I don't think it needs to be threadsafe.